File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if(s=="") return 0;
        int maxi=1;
        unordered_map<char,int> mp;
        int l=0;
        int r=1;
        int n=s.size();
        mp[s[0]]=0;
        while(r<n && l<n){
            if(mp.count(s[r])&& mp[s[r]]>=l){
                l=mp[s[r]]+1;
            }
            maxi=max(maxi,r-l+1);
            mp[s[r]]=r;
            r++;
        }
        return maxi;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n=s.length();
        int l=0;
        HashMap<Character,Integer>hmp= new HashMap<>();
        int maxi=0;
        for(int r=0;r<n;r++){
            char c=s.charAt(r);
            if(hmp.containsKey(c)){
                l=Math.max(l,hmp.get(c)+1);
            }
            maxi=Math.max(maxi, r-l+1);
            hmp.put(c,r);
        }
        return maxi;
    }
}